elastic-agent-client.proto

```diff
@@ -101,26 +129,311 @@ message ActionResponse {
+    string token = 1;
+    // Transaction ID.
+    string txn_id = 2;
+    // Name of the key.
+    string name = 3;
+}
+
+// Response from `DeleteKey`.
+message StoreDeleteKeyResponse {
+    // Empty at the moment, defined for possibility of adding future return values.
+}
+
+// Commits the transaction in the store.
+//
+// Upon error the whole transaction is discarded so no need to call discard after error.
+message StoreCommitRequest {
+    // Token that is used to uniquely identify the connection to the Elastic Agent.
+    string token = 1;
+    // Transaction ID.
+    string txn_id = 2;
+}
+
+// Response from `CommitTxn`.
+message StoreCommitTxnResponse {
+    // Empty at the moment, defined for possibility of adding future return values.
+}
+
+// Discards the transaction in the store.
+message StoreDiscardRequest {
+    // Token that is used to uniquely identify the connection to the Elastic Agent.
+    string token = 1;
+    // Transaction ID.
+    string txn_id = 2;
+}
+
+// Response from `DiscardTxn`.
+message StoreDiscardTxnResponse {
+    // Empty at the moment, defined for possibility of adding future return values.
+}
+
+service ElasticAgentStore {
+    // Key-Value state storage is provided for each unit.
+    //
+    // Transactional store is provided to allow multiple key operations to occur before a commit to ensure consistent
+    // state when multiple keys make up the state of an units persistent state.
+    rpc BeginTxn(StoreBeginTxnRequest) returns (StoreBeginTxnResponse);
+    rpc GetKey(StoreGetKeyRequest) returns (StoreGetKeyResponse);
+    rpc SetKey(StoreSetKeyRequest) returns (StoreSetKeyResponse);
+    rpc DeleteKey(StoreDeleteKeyRequest) returns (StoreDeleteKeyResponse);
+    rpc CommitTxn(StoreCommitRequest) returns (StoreCommitTxnResponse);
+    rpc DiscardTxn(StoreDiscardRequest) returns (StoreDiscardTxnResponse);
+}
+
+// Requests an artifact from the Elastic Agent.
+message ArtifactFetchRequest {
+    // Token that is used to uniquely identify the collection of inputs to the agent. When started this is provided
+    // in the `ConnInfo`.
+    string token = 1;
+    // ID of the artifact.
+    string id = 2;
+    // SHA256 of the artifact.
+    string sha256 = 3;
+}
+
+// Content of the artifact.
+message ArtifactFetchResponse {
+    oneof content_eof {
+        // Artifact content.
+        bytes content = 1;
+        // End-of-file.
+        google.protobuf.Empty eof = 2;
+    }
+}
+
+service ElasticAgentArtifact {
+    // Fetches an artifact from the artifact store.
+    //
+    // Response from this call can be chunked over multiple `ArtifactFetchResponse` for very large responses. A minimum
+    // of two responses will always be returned. The last response has eof set.
+    rpc Fetch(ArtifactFetchRequest) returns (stream ArtifactFetchResponse);
+}
+
+message LogMessage {
+    // ID of the unit.
+    string unit_id = 1;
+    // ECS log message body JSON encoded.
+    bytes message = 2;
```

Comment:
Should we enforce some basic structure already here. For example having timestamp and message already part of the schema and only the rest of the info a blob? This ensures that at least the basic log message is valid.

Reply:
I am unsure here, should we assume that the application uses an ECS format appender to the logs?

Reply:
I think any enforcement here will add overhead. The idea here is that its transparently takes the bytes and just outputs it through stdout.

Reply:
> I think any enforcement here will add overhead.

+1

So would the agent log it in a similar manner to what id does today, wrapping the application's log message on the Agent's own log message? Similar to:

```
message: "Application X log: Whole application's log"
```

Reply:
Let me try to rephrase it:
Do the agent just logs whatever it receives "as is", as plains as `fmt.Println(message)` or does it do or add anything before logging?

Reply:
I think we should consider the log ready to ready or "as is" and don't do any processing on it.